flocker/acceptance/integration/testtools.py

```diff
@@ -3,11 +3,163 @@
 
 
+DOCKER_CLIENT_LOCK = DeferredLock()
+
+
+def deferToThreadWithLock(lock, f, *args, **kwargs):
+    """
+    Like ``deferToThread``, but acquires ``lock`` before calling ``f`` and
+    releases the lock when ``f`` returns or raises.
+    """
+    locking = lock.acquire()
+
+    def locked(lock):
+        return deferToThread(f, *args, **kwargs)
+
+    calling = locking.addCallback(locked)
+
+    def called(result):
+        lock.release()
+        return result
+
+    unlocking = calling.addBoth(called)
+    return unlocking
+
+
+class Container(PClass):
+    """
+    Perform asynchronous docker-py start and remove operations on a Docker
+    container that has been created.
+    Docker-py operations are performed in a threadpool and with a lock in case
+    DockerClient is not thread safe.
+
+    :attr DockerClient client: A DockerClient connected to a specific docker
+        server endpoint.
+    :attr container_id: The unique ID of the container to operate on.
+    """
+    client = field()
+    container_id = field()
+
+    def start(self):
+        """
+        :returns: A ``Deferred`` that fires when the docker API start call
+            completes.
+        """
+        return deferToThreadWithLock(
+            DOCKER_CLIENT_LOCK,
+            self.client.start,
+            container=self.container_id
+        )
+
+    def remove(self):
+        """
+        Forcefully remove the container, even if it is still running.
+
+        :returns: A ``Deferred`` that fires when the docker API remove call
+            completes.
+        """
+        return deferToThreadWithLock(
+            DOCKER_CLIENT_LOCK,
+            self.client.remove_container,
+            self.container_id,
+            force=True,
+        )
+
+
+def create_container(client, create_arguments):
+    """
+    Create a Docker container and return a ``Container`` with which to perform
+    start and remove operations.
+
+    :param DockerClient client: The DockerClient which will be used to create
+        the container.
+    :param dict create_arguments: Keyword arguments to pass to
+        DockerClient.create_container.
+    :returns: A ``Container``.
+    """
+    container_data = client.create_container(**create_arguments)
+    container_id = container_data["Id"]
+    return Container(
+        client=client,
+        container_id=container_id,
+    )
+
+
+def stateful_container_for_test(test, cluster, node, image_name,
+                                dataset, internal_path, internal_port,
+                                external_port):
+    """
+    Create and start a ``Container`` on ``node``.
+    Clean it up when the ``test`` has completed.
+
+    :param TestCase test: The test.
+    :param Cluster cluster: The ``Cluster`` with certificates for
+        authenticating with the docker daemon on ``node``.
+    :param unicode image_name: A name of the Docker image to use.
+    :param Dataset dataset: The mounted Flocker dataset to bind mount into the
```

Comment:
The ``create_dataset(test=self, ...)`` function registers a ``TestCase.addCleanup`` function to remove the dataset using a Flocker API call.